NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -755,6 +755,14 @@ bool CParaNdisTX::DoPendingTasks(CNBL *nblHolder)
 void CNB::MappingDone(PSCATTER_GATHER_LIST SGL)
 {
     m_SGL = SGL;
+#if NDIS_SUPPORT_NDIS650
+    #if ALICLOUD_MAX_FRAGMENTS_IN_ONE_NB
+        if (SGL->NumberOfElements > m_Context->uMaxFragments)
+        {
```

Comment:
Let's say we get SGL->NumberOfElements=62, everything seems OK and we do not use the copy. But the driver adds a fragment for vnet_hdr and also 1 or 2 fragments depending on AnyLayout and VLAN header and the resulting number might be 65. Correct?

Reply:
@ybendito .
Correct, in light of the potential for AllocateAndFillCopySGL to fail, as you mentioned above, I plan to switch to the following code(This is pseudocode, where GetHeaderFragments is used to obtain the number of fragments contained in the header information). Move the logic of AllocateAndFillCopySGL into FillDescriptorSGList. What do you think?

bool CNB::FillDescriptorSGList(CTXDescriptor &Descriptor, ULONG ParsedHeadersLength) const
{
#if MAX_FRAGMENTS_IN_ONE_NB
    if (m_SGL->NumberOfElements > ALICLOUD_MAX_FRAGMENTS_IN_ONE_NB - Descriptor.GetHeaderFragments())
    {
        if (!AllocateAndFillCopySGL())
        {
            return false;
        }
        else
        {
            return Descriptor.SetupHeaders(ParsedHeadersLength) &&
                MapCopyDataToVirtioSGL(Descriptor, ParsedHeadersLength);
        }
    }
#endif
    return Descriptor.SetupHeaders(ParsedHeadersLength) &&
        MapDataToVirtioSGL(Descriptor, ParsedHeadersLength + NET_BUFFER_DATA_OFFSET(m_NB));
}